rfcs/0156-no-direct-pushes.md

```diff
@@ -0,0 +1,173 @@
+
+</details>
+
+## Emergency changes
+
+Sometimes channels have blocking breakages and need to be fixed as soon as possible (citation needed).
+Currently this can be done with direct pushes, but a pull request will be required with this proposal.
+The time required to fix such breakages however is barely affected: Since there is currently no requirement for pull requests to be approved or pass CI, they can get merged immediately after opening if necessary.
+
+## Staging workflow
+
+The staging workflow is not affected because it [already uses pull requests](https://github.com/NixOS/nixpkgs/pull/241951) for all merges into the affected branches.
+
+## Direct push detection
+
+There is a [GitHub Actions workflow to detect directly pushed commits](https://github.com/NixOS/nixpkgs/blob/0b411c1e040870e89a3e598437e708979137b665/.github/workflows/direct-push.yml).
+When detected, it creates a comment in the commit pointing out that this is discouraged and linking to [this issue](https://github.com/NixOS/nixpkgs/issues/118661), where it's easy to see all direct pushes.
+The script occasionally has false positives ([issue](https://github.com/NixOS/nixpkgs/issues/240314), which creates some unnecessary commit comment notifications.
+
+With this proposal accepted it won't be possible to directly push commits anymore, making that workflow unnecessary.
+It can be removed and the above two issues can be closed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Even trivial changes that don't need a review now require a pull request, which takes more effort and creates noise for reviewers.
+- It takes slightly longer to push a quick fix in emergency situations.
+
+# Alternatives
+[alternatives]: #alternatives
+
+# Prior art
+[prior-art]: #prior-art
+
+The previous [RFC 79](https://github.com/NixOS/rfcs/pull/79) attempted to do the same, but:
+- It had a mistaken estimate for the percentage of direct master commits, calculating it to be 46.85% in the last year.
+  The mistake was assuming that all non-merge commits were direct pushes.
+  This made it seem like the change was much more impactful than it actually would've been.
+- It was too ambitious by also proposing to require accepting reviews for all pull requests.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+# Future work
+[future]: #future-work
+More restrictions could be implemented in the future:
+- Require all pull requests to have at least one approval, therefore preventing people from merging their own pull requests (it's not possible to approve your own pull request)
+- Require an approval by a code owner, properly establishing code ownership over all parts of Nixpkgs
+- Require CI to pass, maybe using GitHub's new [merge queue's](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue)
```

Comment:
Given that
-  in the long term Nixpkgs should be ported to many platforms
- currently the Darwin builds are messy

I think this CI quest needs more polishing.

Reply:
Well, it's a possible future direction, serving more to mark things «out of scope though related»